src/hotspot/share/gc/shared/gcLogPrecious.cpp

```diff
@@ -23,15 +23,48 @@
 
 #include "precompiled.hpp"
 #include "gc/shared/gcLogPrecious.hpp"
+#include "runtime/atomic.hpp"
 #include "runtime/mutex.hpp"
 #include "runtime/mutexLocker.hpp"
+#include "runtime/os.hpp"
 
-stringStream* GCLogPrecious::_lines = NULL;
 stringStream* GCLogPrecious::_temp = NULL;
 Mutex* GCLogPrecious::_lock = NULL;
 
+class GCLogPreciousLine : public CHeapObj<mtGC> {
+  const char* const _line;
+  GCLogPreciousLine* volatile _next;
+
+public:
+  GCLogPreciousLine(const char* line) :
+      _line(line),
+      _next(NULL) {}
+
+  const char* line() const { return _line; }
+
+  GCLogPreciousLine* next() const { return Atomic::load_acquire(&_next); }
+  void set_next(GCLogPreciousLine* next) { Atomic::release_store(&_next, next); }
+};
+
+GCLogPreciousLine* volatile GCLogPrecious::_head = NULL;
+GCLogPreciousLine* GCLogPrecious::_tail = NULL;
+
+void GCLogPrecious::append_line(const char* const str) {
+  GCLogPreciousLine* line = new GCLogPreciousLine(str);
+
+  GCLogPreciousLine* head = _head;
+  GCLogPreciousLine* tail = _tail;
+
+  if (head == NULL) {
+    Atomic::release_store(&_head, line);
+  }
+  if (_tail != NULL) {
+    tail->set_next(line);
+  }
```

Comment:
Either 61 should be using `tail` or 62 should be using `_tail`.  If the latter then the `tail` local variable is unused.